### ISSUE TITLE
reportTypedDictNotRequiredAccess seems not works correctly when total=False

### ISSUE THREAD

**Issue by Diogo-Rossi** [microsoft/pylance-release#2256]
<h3> Environment data </h3>

-   Language Server version: 2022.1.1
-   OS and version: win32 x64 - Windows 10.0.17763.1935
-   Python version: Python 3.9.7 (default, Sep 16 2021, 16:59:28) [MSC v.1916 64 bit (AMD64)] :: Anaconda, Inc. on win32
-   python.analysis.typeCheckingMode: basic

<h3> Expected behaviour </h3>

PyLance should allow assign values to keys of the specified type in `TypedDict`

<h3> Actual behaviour </h3>

PyLance trigger the `reportTypedDictNotRequiredAccess` when assigned keys are not mandatory (i.e., not required, with `total=False`). It should not trigger this error, like it does when `total=False` is not present.

____
<h3> Code Snippet / Additional information </h3>

Using the same example as used in the related issue #1926

```python
from typing import TypedDict

class MyDict(TypedDict, total=False):
    key1: int
    key2: str

my_dict: MyDict = {
    'key1': 1,
    'key2': "test"
}

my_dict['key1'] = 2  # Here there is an "error"
```

Removing `total=False`, the error doesn't show
```python
from typing import TypedDict

class MyDict(TypedDict):
    key1: int
    key2: str

my_dict: MyDict = {
    'key1': 1,
    'key2': "test"
}

my_dict['key1'] = 2  # No "error" anymore
```
___
<h3> Screen captures </h3>

No error:
![image](https://user-images.githubusercontent.com/57204023/149804381-3dcf718e-1343-4a3d-a9dd-bc9e65177de7.png)

Error:
![image](https://user-images.githubusercontent.com/57204023/149799967-37f2a00c-ad3f-4e0f-97a6-3e95267ab6b0.png)

View problem:
![image](https://user-images.githubusercontent.com/57204023/149800012-0b767594-595b-4fcb-a874-b08156db88b9.png)

**Comment by erictraut** [comment]
Thanks for the bug report. There were two bugs here:
1. This shouldn't have been reported as a `reportTypeDictNotRequiredAccess` diagnostic. Those should be reported only for "get" accesses, not "set" accesses. The correct diagnostic category here is `reportGeneralTypeIssues`.
2. The access was using the _narrowed type_ of the field based on its previous assignment of `1`, but it should have been using the _declared type_ of the field (in this case `int`). Using the proper type here, no error is produced.

I've addressed both issues and beefed up our tests to cover this case. The fix will be included in the next pylance release.

**Comment by Diogo-Rossi** [comment]
@erictraut Thanks.

To add information (that may improve the discussion):
I've notice that defining the object (assigning the keys) by the *TypedDict constructor* (rather than by the Python dict expression syntax `{}`) does not show error.

```python
from typing import TypedDict

class MyDict(TypedDict, total=False):
    key1: int
    key2: str

# direct TypedDict constructor
my_dict = MyDict(
    key1 = 1,
    key2 = "test"
)

my_dict['key1'] = 2   # No "error" anymore
```

**Comment by bschnurr** [comment]
This issue has been fixed in version 2022.1.3, which we've just released. You can find the changelog here:  [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202213-21-january-2022)
